### PR TITLE
On-demandify associated item retrieval

### PR DESCRIPTION
**Pull request by cramertj**
Part of #40614.

I also started converting `adt_def`, but I decided to open a PR with just this bit first to make sure I'm going about this correctly.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/ty/mod.rs [hunk 1]
```diff
@@ -2623,3 +2575,47 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
...
+{
+    let id = tcx.hir.as_local_node_id(def_id).unwrap();
+    let parent_id = tcx.hir.get_parent(id);
+    let parent_def_id = tcx.hir.local_def_id(parent_id);
+    let parent_item = tcx.hir.expect_item(parent_id);
+    match parent_item.node {
+        hir::ItemImpl(.., ref impl_trait_ref, _, ref impl_item_refs) => {
+            for impl_item_ref in impl_item_refs {
+                let assoc_item =
+                    tcx.associated_item_from_impl_item_ref(parent_def_id,
```
Comment (nikomatsakis) [inline src/librustc/ty/mod.rs]: This looks basically right. I would probably rewrite this `for` loop (and the one below) to something like:

```rust
if let Some(impl_item_ref) = impl_item_refs.iter().find(|i| i.id.node_id == id) {
    return tcx.associated_item_from_impl_item_ref(...); // same parameters you have already
}
```

Comment (nikomatsakis) [inline src/librustc/ty/mod.rs]: maybe with an `assert!` that `assoc_item.def_id == def_id` at the end =)

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -2623,3 +2575,47 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         }
     }
 }
+
+fn associated_item<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
```
Comment (eddyb) [inline src/librustc/ty/mod.rs]: I think this should be in `rustc::hir`, including `associated_item_from_{impl,trait}_item_ref`.

Comment (nikomatsakis) [inline src/librustc/ty/mod.rs]: I opened up https://github.com/rust-lang/rust/issues/40697 where we can discuss the precise plan. It feels a bit like this should go into a follow-up PR to me; this PR is just converting to on-demand but keeping the existing structure.

**Pull request merged**